src/patchenv.jl

```diff
@@ -0,0 +1,64 @@
+struct PatchEnvName{id} <: Cassette.AbstractContextName end
```

Comment:
Hmm. What's the reason for manually implementing a `<:Cassette.AbstractContextName` here? Cassette doesn't expect users to do that. Could be fine, but it's e.g. an implementation detail that has license to break at any time (not that Cassette is super stable in it's pre-1.0 stage anyway 😛).

Reply:
Actually, this is almost certainly problematic now that I think about it. `@context` does way more than just define context names, it also makes sure the method table is setup correctly for your new context type. Why skip using it?

Reply:
I ended up proposing this as I was trying to avoid an extra `eval` call

Reply:
See: https://github.com/invenia/Mocking.jl/pull/60#discussion_r240357000

Reply:
Taking a look at what `@context` does.
https://github.com/jrevels/Cassette.jl/blob/3a858bd9d8d167ab8708903f3f4f29961d7a5b1a/src/context.jl#L221

Is there a reason most of those function definitions can't be made to run on the abstract type?

Reply:
Ok, gotcha.

So we could do something like

```
@context MockingContext

function PatchEnv()
    ctx_metadata = Val{gensym()}() # This is for dispatch purposes 
    ctx = MockingCtx(metadata=ctx_metadata)
    return PatchEnv{typeof(ctx)}(ctx)
end
```

And different `PatchEnv`s would generate different contexts for dispatch purposes when we generate the `execute`s,
because of the fact that the type of the metadata is a typeparam of the context type.

Reply:
Yup.  Naively, I'd assume it'd be fine to just make it `PatchEnvContext`, instead of having the extra wrapping layer, but maybe you'd want to split them out for other reasons that I haven't seen yet (since I've only just skimmed this PR so far).

Reply:
Each `PatchEnv`  has a different set of patches to be applied.
Which translates to needing to declare a different set of `execute` methods.